src/utils/timer.ts

```diff
@@ -0,0 +1,105 @@
+import {
+  addTimedelta,
+  TimeComponents,
+  toMilliseconds
+} from "./time"
+
+export type TimerComponentsParameters = {
+  duration: TimeComponents,
+  tick?: TimeComponents,
+  continues?: boolean,
+  onTick?: (timer: Timer) => void,
+  onComplete: (timer: Timer) => void,
+  onStart?: (timer: Timer) => void,
+}
+
+/**
+ * Generic timer implementation
+ * @author Mike Reiche <mike@reiche.world>
+ */
+export class Timer {
+  private readonly _params: TimerComponentsParameters
+  private _tickInterval: number
+  private _durationTimeout: number
+  private _startDate: Date
+  private _durationMs: number
+  private _completeDate: Date
+
+  constructor(params: TimerComponentsParameters) {
+    this._params = params
+    this._startDate = new Date()
+    this._completeDate = this._startDate
+    this._updateDuration()
+  }
+
+  /**
+   * (Re)starts the timer
+   */
+  start() {
+    this.dispose()
+    if (this._params.tick && this._params.onTick) {
+      this._tickInterval = window.setInterval(() => this._params.onTick(this), toMilliseconds(this._params.tick))
+    }
+    this._startDate = new Date()
+    this._completeDate = addTimedelta(this._startDate, this._params.duration)
+    this._updateDuration()
+    this._durationTimeout = window.setTimeout(() => this._onComplete(), this._durationMs)
```

Comment:
Using two different timers will mean that we may call `onComplete` before the next tick finishes. While technically better than common clock implementations, I am unsure whether this will not be surprising behaviour - most clock-based timers will always complete _exactly on_ a tick. It would also seem to be "technically a bug": If the remaining time ticks down at the defined intervals, "of course" the timer should only finish, once it reaches/crosses zero - and this can only happen at a tick.

... Nevertheless I do not really have a strong opinion whether this is fine or bad.